File: janMarkusMilan/quiz/q1.js
// Given an array of positive numbers and a positive number ‘k,’
// find the maximum sum of any contiguous subarray of size ‘k’.

/*
Input: [3, 1, 2, 4], 2
Output: 6
Explanation: Since the size for k is 2, we are adding two numbers together at most.
             Looking at the array, the contiguous subarray that gives us the max is [2, 4].

-----

Analyzing the problem:
    1. Contiguous subarray means a portion in the array that is connected
    2. The maximum size of the "window" is determined by the input 'k'
    3. We are only given positive numbers

Possible constraints:
    1. The input 'k' should be a valid number (will be in the bounds of the array length)
    2. The input 'k' will always be a positive number
    3. The array will only contain positive numbers

Edge cases:
    1. If the array is empty, return 0

Approaching the problem:
    1. We can use the sliding window technique to form a "window" with a specific length over a portion of the array.
       To do so, we initialize two pointers starting at 0 and increment conditionally.
    2. We need to find the maximum sum, meaning we add the numbers in the "window" and take the max.
       So, we probably need a sum and a max variable as well.
    3. Loop through the array to get each value, and add it to our sum.
    4. Once we have the desired window length:
        - Take the max
        - Increment our pointers by 1, meaning we slide the window over
        - Decrement the sum by the left-most value
    5. At the very end, we can return the max
*/

const maxSubArray = function (nums, k) {
  let sum = 0;
  let max = 0;

  let leftPtr = 0;
  let rightPtr = 0;
  while (rightPtr < nums.length) {
    sum += nums[rightPtr];

    if (rightPtr >= k - 1) {
      max = Math.max(max, sum);
      sum -= nums[leftPtr];
      leftPtr++;
    }

    rightPtr++;
  }

  return max;
};

console.log(maxSubArray([3, 1, 2, 4], 2)); // 6 -> [2, 4]
console.log(maxSubArray([1, 9, 3, 5, 2, 7, 2], 4)); // 19 -> [9, 3, 5, 2]
console.log(maxSubArray([], 0)); // 0 -> []

File: mathewDavidov/quiz/q1.java
/**
 * Problem Statement:
 * Given an array of positive numbers and a positive number ‘k,’ 
 * find the maximum sum of any contiguous subarray of size ‘k’.
 * 
 * Example:
 * [5, 1, 4, 6, 2], k = 2
 * Output: 10
 * 
 * Edge cases:
 * k > nums.length -> return -1
 * null / empty array -> return -1
 * 
 * Sliding window
 * initialize a maxSum and currentSum
 * initialize a windowStart
 * loop from 0 to nums.length:
 *   add the current index to currentSum
 *   if we have a window of size k:
 *     update maxSum to be the max of maxSum and currentSum
 *     move the window forward by subtracting from the windowStart
 * return the maxSum
 * 
 * Runtime:
 * Time: O(n)
 * Space: O(1)
 */

public class q1 {
    public static int maxSumSub(int[] nums, int k) {
        if (nums == null || nums.length == 0 || k > nums.length) {
            return -1;
        }

        int maxSum = -1, currentSum = 0, windowStart = 0;

        for (int windowEnd = 0; windowEnd < nums.length; windowEnd++) {
            currentSum += nums[windowEnd];

            if (windowEnd >= k - 1) {
                maxSum = Math.max(maxSum, currentSum);
                currentSum -= nums[windowStart++];
            }
        }

        return maxSum;
    }

    public static void main(String[] args) {
        System.out.println(maxSumSub(new int[]{5, 1, 4, 6, 2}, 2));
        System.out.println(maxSumSub(new int[]{3, 5, 2, 7, 2, 4, 6}, 4));
        System.out.println(maxSumSub(new int[]{3, 5, 2, 7, 2, 4, 6}, 7));
        System.out.println(maxSumSub(new int[]{}, 2));
        System.out.println(maxSumSub(null, 1));
        System.out.println(maxSumSub(new int[]{5, 1, 4, 6, 2}, 10));
    }
}